qadence/backends/adjoint.py

```diff
@@ -0,0 +1,159 @@
+        state = gate(state)
+
+    projected_state = o(state)  # Apply the observable to the state.
+    expval = overlap(state, projected_state)  # Compute the expected value.
+
+    2. Backward pass:
+    grads = []
+    for gate in reversed(c.gates): # Iterate through c.gates in "reverse", so gateN, gateN-1 etc.
+        state = dagger(gate)(state)  # 'Undo' the gate by applying its dagger.
+        if gate is Parametric:
+            mu = jacobian(gate)(state) # Compute the jacobian of the gate w.r.t its parameter.
+            grads.append(2 * overlap(mu, projected_state) # Compute the gradient.
+        projected_state = dagger(gate)(projected_state)  # 'Undo' the gate from the projected_state.
+
+    Current Limitations:
+
+    (1) The adjoint method is only available in the pyqtorch backend.
+    (2) Parametric observables are not supported.
+    (3) Multiple observables are not supported.
+    (4) Higher order derivatives are not natively supported.
+    (5) Only expectation values can be differentiated, not wave functions.
+    """
+
+    @staticmethod
+    @no_grad()
+    def forward(
+        ctx: Any,
+        circuit: PyQCircuit,
+        observable: PyQCircuit,
+        state: Tensor,
+        param_names: list[str],
+        *param_values: Tensor,
+    ) -> Tensor:
+        for param in param_values:
+            param = param.detach()
+        ctx.circuit = circuit
+        ctx.observable = observable
+        ctx.param_names = param_names
+        values = param_dict(param_names, param_values)
+        ctx.out_state = circuit.run(state, values)
+        ctx.projected_state = observable.run(ctx.out_state, values)
+        ctx.save_for_backward(*param_values)
+        return overlap(ctx.out_state, ctx.projected_state)
+
+    @staticmethod
+    @no_grad()
+    def backward(ctx: Any, grad_out: Tensor) -> tuple:
+        param_values = ctx.saved_tensors
+        values = param_dict(ctx.param_names, param_values)
+
+        def _apply_adjoint(ctx: Any, op: Module) -> list:
+            grads: list = []
+            if isinstance(op, PyQHamiltonianEvolution):
+                generator = op.block.generator
+                time_param = values[op.param_names[0]]
+                ctx.out_state = apply_operator(ctx.out_state, op.dagger(values), op.qubit_support)
+                # A HamEvo can have a parametrized (1) time evolution and/or (2) generator.
+                if (
+                    isinstance(generator, AbstractBlock)
+                    and generator.is_parametric
+                    and values[op.param_names[1]].requires_grad
+                ):
+                    # If the generator contains a trainable parameter, we compute its gradient.
+                    mu = apply_operator(
+                        ctx.out_state, op.jacobian_generator(values), op.qubit_support
+                    )
+                    grads.append(2 * overlap(ctx.projected_state, mu))
+                elif time_param.requires_grad:
```

Comment:
Could there be a case where the HamEvo contains both a trainable parameter and time ?

Reply:
yes, thats possible, i think we are handling that?

Reply:
good catch! we dont. added